Add NuRaft handler commit_config()

AlexRamRam opened the pull request:
Add NuRaft handler commit_config() to maintain synchronization of last committed index with latest log index

Fix #125
Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

AlexRamRam committed:
Add NuRaft handler commit_config()

Maintain synchronization of last committed index with latest log index

Fix #125
Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

metalicjames commented on src/uhs/atomizer/atomizer/state_machine.cpp:
```diff
@@ -99,6 +99,12 @@ namespace cbdc::atomizer {
+    void state_machine::commit_config(
+        const nuraft::ulong log_idx,
+        nuraft::ptr<nuraft::cluster_config>& /*new_conf*/) {
+        m_last_committed_idx = log_idx;
```
Is it worth asserting here that `log_idx > m_last_committed_idx`, or even stronger that `log_idx == m_last_committed_idx + 1`?

AlexRamRam replied:
I agree about `log_idx == m_last_committed_idx + 1`. Will add this to both `commit()` and `commit_config()`

metalicjames commented on src/uhs/atomizer/atomizer/state_machine.hpp:
```diff
@@ -45,6 +45,12 @@ namespace cbdc::atomizer {
 
+        /// Handler for the raft cluster configuration changes
+        /// \param log_idx Raft log number of the configuration change.
+        void commit_config(
```
I think we should add this to all the state machines, not just atomizer.

AlexRamRam replied:
Agreed.

AlexRamRam committed:
Add Nuraft handler commit_config() for all state_machine classes

And, assert in all state_machine commit() and commit_config() hooks that incoming log_idx is one higher than last committed index

Signed-off-by: Alexander Jung <104335080+AlexRamRam@users.noreply.github.com>

HalosGhost reviewed (approved):
This looks like a very reasonable solution to the problem. Both minimal and clear.

Looks good to me.
